### PR TITLE
Update Reader interfaces

### PR DESCRIPTION
**Pull request by roxblnfk**
| Q             | A
| ------------- | ---
| Breaks BC?    | ✔️
| Fixed issues  | #47 

- Limit in the IterableDataReader is not set by default
- Added `@template*` and `@psalm-immutable` docblocks into reader interfaces
- Added `self` return-types for immutable methods of reader interfaces

---

Fix broken BC in other reposiitories:

- [x] `yii-cycle` -- `SelectDataReader`

### CODE REVIEW

File: src/Reader/Iterable/IterableDataReader.php [hunk 1]
```diff
@@ -27,7 +27,7 @@ class IterableDataReader implements DataReaderInterface
     protected iterable $data;
     private ?Sort $sort = null;
     private ?FilterInterface $filter = null;
-    private int $limit = self::DEFAULT_LIMIT;
+    private ?int $limit = null;
```
Comment (xepozz) [inline src/Reader/Iterable/IterableDataReader.php:30]: I think would be nice set `-1` as a "unlimited value"

Comment (roxblnfk) [reply src/Reader/Iterable/IterableDataReader.php:30]: By the way, if the restriction is set as 0, then it is not clear what behavior should be expected from the reader.

I think we need to assert this behavior in the interface comment. Perhaps `withLimit(0)` will mean unlimited and `withLimit(-1)` will throw exception.

Comment (xepozz) [reply src/Reader/Iterable/IterableDataReader.php:30]: Ok let's do it

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
👍

**Pull request merged**